File: elixir/61-RotateList/main.exs
defmodule ListNode do
  @type t :: %__MODULE__{
          val: integer,
          next: ListNode.t() | nil
        }
  defstruct val: 0, next: nil
end

defmodule Solution do
  defp list_node_length(nil),  do: 0
  defp list_node_length(node), do: 1 + list_node_length(node.next)

  def rotate_right(head, 0), do: head
  def rotate_right(nil, _),  do: nil

  @spec rotate_right(head :: ListNode.t | nil, k :: integer) :: ListNode.t | nil
  def rotate_right(head, k) do
    length = list_node_length(head)
    shift = rem(k, length)

    new_tail = copy(head, length - shift, nil)
    new_head = skip(head, length - shift)
    copy(new_head, shift, new_tail)
  end

  defp copy(_, 0, tail),        do: tail
  defp copy(node, count, tail), do: %ListNode{val: node.val, next: copy(node.next, count - 1, tail)}

  defp skip(node, 0),     do: node
  defp skip(node, count), do: skip(node.next, count - 1)
end


defmodule Tests do
  defp list_to_string(list) when list == nil do
    ""
  end

  defp list_to_string(list) do
    "#{list.val} " <> list_to_string(list.next)
  end

  def example1 do
    list = %ListNode{val: 1, next: %ListNode{val: 2, next: %ListNode{val: 3, next: %ListNode{val: 4, next: %ListNode{val: 5}}}}}
    IO.puts(list_to_string(Solution.rotate_right(list, 2)))
  end

  def example2 do
    list = %ListNode{val: 0, next: %ListNode{val: 1, next: %ListNode{val: 2}}}
    IO.puts(list_to_string(Solution.rotate_right(list, 4)))
  end

  def example3 do
    list = %ListNode{val: 0, next: %ListNode{val: 1, next: %ListNode{val: 2}}}
    IO.puts(list_to_string(Solution.rotate_right(list, 0)))
  end
end

Tests.example1
Tests.example2
Tests.example3

File: python/61-RotateList/main.py
from typing import Optional
from typing import Tuple

import unittest

import sys
sys.path.append('../modules')
from leetcode import ListNode


class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 0:
            return head

        (length, tail) = self.length_and_tail(head)
        if length == 0 or length == 1:
                return head
        
        # k can be larger than the length of the list so we should work
        # out how much we actually do heed to shift by
        shift = k % length
        if shift == 0:
            return head

        to_skip = length - shift

        new_head = head
        previous = None

        while to_skip != 0:
             previous = new_head
             new_head = new_head.next
             to_skip -= 1

        previous.next = None
        tail.next = head

        return new_head
        
    
    def length_and_tail(self, head: Optional[ListNode]) -> Tuple[int, Optional[ListNode]]:
        count = 0
        tail = None

        while head is not None:
            count += 1
            tail = head
            head = head.next

        return (count, tail)
    
class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        l = ListNode.makeList(0, 1, 2)
        result = solution.rotateRight(l, 1)
        self.assertListEqual(ListNode.flatten(result), [2, 0, 1])

    def test_example2(self):
        solution = Solution()
        l = ListNode(0, ListNode(1, ListNode(2)))
        solution.rotateRight(l, 6)

    def test_example3(self):
        solution = Solution()
        l = ListNode(1)
        solution.rotateRight(l, 1)

    def test_example4(self):
        solution = Solution()
        l = ListNode(1)
        solution.rotateRight(l, 99)


if __name__ == '__main__':
    unittest.main()